interoperator/controllers/schedulers/sflabelselectorscheduler/sflabelselectorscheduler_controller.go

```diff
@@ -0,0 +1,245 @@
+
+	sfNamespace := os.Getenv(constants.NamespaceEnvKey)
+	if sfNamespace == "" {
+		sfNamespace = constants.DefaultServiceFabrikNamespace
+	}
+	plan := &osbv1alpha1.SFPlan{}
+	namespacedName := types.NamespacedName{
+		Name:      sfServiceInstance.Spec.PlanID,
+		Namespace: sfNamespace,
+	}
+	err := r.Get(ctx, namespacedName, plan)
+	if err != nil {
+		return "", err
+	}
+	service := &osbv1alpha1.SFService{}
+	namespacedName = types.NamespacedName{
+		Name:      sfServiceInstance.Spec.ServiceID,
+		Namespace: sfNamespace,
+	}
+	err = r.Get(ctx, namespacedName, service)
+	if err != nil {
+		return "", err
+	}
+
+	labelSelectorTemplate, err := plan.GetTemplate(osbv1alpha1.ClusterLabelSelectorAction)
+	if err != nil || labelSelectorTemplate.Type != "gotemplate" {
+		log.Info("plan does not have clusterlabel template")
+		return "", err
+	}
+
+	content := labelSelectorTemplate.Content
+	values := make(map[string]interface{})
+	instanceObj, err := dynamic.ObjectToMapInterface(sfServiceInstance)
+	values["instance"] = instanceObj
+	if err != nil {
+		return "", err
+	}
+
+	if service != nil {
+		serviceObj, err := dynamic.ObjectToMapInterface(service)
+		values["service"] = serviceObj
+		if err != nil {
+			return "", err
+		}
+	}
+	if plan != nil {
+		planObj, err := dynamic.ObjectToMapInterface(plan)
+		values["plan"] = planObj
+		if err != nil {
+			return "", err
+		}
+	}
+
+	renderer, err := gotemplate.New()
+	if err != nil {
+		return "", err
+	}
+	rendererInput := gotemplate.NewInput("", content, "test", values)
+	rendererOutput, err := renderer.Render(rendererInput)
+	if err != nil {
+		return "", err
+	}
+	labelSelector, err := rendererOutput.FileContent("main")
+	if err != nil {
+		return "", err
+	}
+	return labelSelector, nil
+}
+
+func (r *SFLabelSelectorScheduler) schedule(labelSelector string) (string, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("labelSelector", labelSelector)
+	labelSelector = strings.TrimSuffix(labelSelector, "\n")
+	label, _ := labels.Parse(labelSelector)
+	log.Info("Parsed Label is: ", "label", label)
+	clusters, err := r.clusterRegistry.ListClusters(&client.ListOptions{
+		LabelSelector: label,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	log.Info("Cluster size is", "length", len(clusters.Items))
+	if len(clusters.Items) == 0 {
+		log.Info("no cluster matching the criteria, returning failure")
+		return "", errors.NewSchedulerFailed(constants.LabelSelectorSchedulerType, "No clusters found with matching criteria: "+labelSelector, nil)
+	}
+	if len(clusters.Items) == 1 {
+		log.Info("Only one cluster matching the criteria", "cluster name", clusters.Items[0].GetName())
+		return clusters.Items[0].GetName(), nil
+	}
+
+	sfserviceinstances := &osbv1alpha1.SFServiceInstanceList{}
+	err = r.List(ctx, sfserviceinstances, &client.ListOptions{})
+	if err != nil {
+		log.Error(err, "failed to list all sfserviceinstances")
+		return "", err
+	}
+
+	counts := make(map[string]int64)
```

Comment:
Was there a reason for using 64 bit values and not 32 bit ones ? Could we ever run into a situation where we have more than ~4 billion instances ?